config/env_spec_request.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package config defines the API Runtime Control config and provides
+// the config loading and validation functions.
+package config
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/apigee/apigee-remote-service-golib/v2/auth/jwt"
+	envoy "github.com/envoyproxy/go-control-plane/envoy/service/auth/v3"
+)
+
+// NewEnvironmentSpecRequest creates a new EnvironmentSpecRequest
+func (e *EnvironmentSpecExt) NewEnvironmentSpecRequest(req *envoy.CheckRequest) *EnvironmentSpecRequest {
+	return &EnvironmentSpecRequest{
+		EnvironmentSpecExt: e,
+		request:            req,
+	}
+}
+
+// EnvironmentSpecRequest extends a request to support operations within an EnvironmentSpec
+type EnvironmentSpecRequest struct {
+	*EnvironmentSpecExt
+	jwtRequirements map[string]*JWTAuthentication // JWTAuthentication.Name ->
+	request         *envoy.CheckRequest
+	jwtResults      map[string]*jwtResult // JWTAuthentication.Name ->
+	verifier        jwt.Verifier
+	apiSpec         *APISpec
+	operation       *APIOperation
+}
+
+type jwtClaims map[string]interface{}
+
+type jwtResult struct {
+	claims jwtClaims
+	err    error
+}
+
+// GetAPI uses the base path to return an APISpec
+func (e *EnvironmentSpecRequest) GetAPISpec() *APISpec {
```

Comment:
nil check?

Reply:
I started to add the checks, but it's never valid to call these methods on a nil struct and would only happen because of programmer error, not user imput, thus I believe it's better to not pretend the call is valid.

Reply:
I fully agree. It just feels safer to me to reduce the possibility of triggering a panic ;-)

Reply:
It does... but also, It could hide error conditions in the code and cause improper behavior.

Reply:
You could add error to the return values, in which case error checking would have to be done by the caller...